### ISSUE TITLE
Document environment history stacking behavior

### ISSUE THREAD

**Issue by zmbc** [conda/conda#14200]
### Checklist

- [X] I added a descriptive title
- [X] I searched open reports and couldn't find a duplicate

### What happened?

As a user, I'm confused about the status of https://github.com/conda/conda/issues/7474 -- is it a bug or intended behavior? I would like to turn off "environment history stacking" and cannot find any way to do it, nor is it clear whether it is supposed to be on by default. As @tjd2002 said in that issue:

> I wonder if the desired behavior I'm describing may indeed have been the original intent of the changes? After all, issue https://github.com/conda/conda/issues/7194 is titled "In conda 4.6, change default value of max_shlvl to 0" (i.e. disable both environment PATH stacking and environment history stacking?).

**Comment by travishathaway** [comment]
@zmbc,

Environment stacking behavior is currently documented here:

- https://docs.conda.io/projects/conda/en/stable/user-guide/tasks/manage-environments.html#nested-activation

If we add anything to the documentation, it should be there. While the `--stack` and `auto_stack` settings are explained there, the "shell level" is not. There is currently no way to turn this off. If you would like to that to be added as an ability in conda, it should be added as a separate feature request.

**Comment by zmbc** [comment]
Thanks @travishathaway. Yes, the "environment stacking" behavior is documented, but "environment history stacking" (perhaps not the best term, I got this from #7474) is not.

What I'd like to understand is whether this is a bug or missing documentation. As noted in #7474, conda _used_ to have the ability to turn off or limit the depth of the "history stack," and the title of #7194 seems to imply that the intended proposal was to make it off by default (like actual environment stacking).

Which of these would be more likely to be accepted?
- A PR to reintroduce a maximum on the history stack, and set the default for that maximum to 0.
- A documentation PR explaining the history stack. It may be just me, but I find the current behavior incredibly confusing and therefore quite difficult to document clearly.

**Comment by zmbc** [comment]
The PR (#7195) that removed the `max_shlvl` configuration parameter says:

>To be clear, part of this PR is also removing the max_shlvl configuration parameter. It's no longer needed. It was originally introduced to keep users from unwittingly stacking many levels of environments beyond two.

This was not the only effect of the max_shlvl configuration parameter, and I wonder if removing it was an oversight.

**Comment by travishathaway** [comment]
@zmbc,

The most likely PR that would be accepted would be one about documenting current behavior.

If one were submitted that either adds a new setting or changes current behavior, it will have to be more closely scrutinized, and it is not a given that we would accept it.

**Comment by zmbc** [comment]
@travishathaway I understand that behavior change will necessarily require more review; that wasn't quite my question. My question is whether this behavior is intended or a bug, and therefore which change is correct in the long run. Given that nobody has jumped in here to say this behavior was intended, I am beginning to think it is a bug.

**Comment by tjd2002** [comment]
I think the discussion at #7194 makes it clear that the current behavior is what was intended by #7195, and not a bug.

At #7194, @kalefranz (the author of #7195) says: 
> The ideal situation is that, while a stack of active environments should be available to push onto and pop off of, only a single environment should be able to be active at any given time.

And then later says:
> My preference would be to have `push` and `pop` just be the `activate` and `deactivate` we already have

Further when he originally documented the (now removed) `max_shlvl` config item at #6892, he described it as:
"The maximum number of stacked active conda environments." I.e. it was (at least primarily) intended to limit the depth of the stack of active envs, not the depth of the history stack.

(If you do decide to propose a change in behavior, I think I'd favor it: I don't see the benefit of maintaining a history stack *when not stacking active envs*, and I think it would be more intuitive to have `activate` simply switch to the new env in a stateless way. `conda activate` with no env specified could still be used as a shortcut for `conda activate base`, and `deactivate` would come closer to doing what it says in plain English. Getting rid of the history stack (outside of env stacking) would also have the benefit of getting rid of some hard-to-reason-about behavior, when mixing a series of calls to `activate` with and without `--stack`)

**Comment by zmbc** [comment]
Thanks for that digging @tjd2002 -- I think you're right. I'll open a new issue to request a behavior change.

**Comment by zmbc** [comment]
I'd like to keep this issue open, as the current behavior is still not documented AFAICT.
